### PR TITLE
Add type label into headless/normal services

### PR DESCRIPTION
**Pull request by averdagu**
Adding type label to easier search between headless or normal service and use them to filter only the normal services during scale-down service removal.

Resolves: [OSPRH-4120](https://issues.redhat.com//browse/OSPRH-4120)

### CODE REVIEW

File: controllers/ovndbcluster_controller.go [hunk 1]
```diff
@@ -581,11 +586,15 @@ func (r *OVNDBClusterReconciler) reconcileServices(
 	}
 
 	// Delete any extra services left after scale down
+	normalServiceLabels := map[string]string{
+		common.AppSelector: serviceName,
+		"type":             "normal",
```
Comment (averdagu) [inline controllers/ovndbcluster_controller.go:591]: Not a big fan of the "normal" name, but the other name that I thought was "service" but can be "confusing" since both Headless and "normal" service are both services. I'm open to new namings

**Comment by lewisdenny** [comment]
Hey @averdagu I started looking into this out of curiosity but didn't get time to finish, just sharing my findings: 
[1] 
```
  - lastTransitionTime: "2024-02-02T15:32:37Z"
    message: OpenStackControlPlane Neutron in progress
    reason: Requested
    severity: Info
    status: "False"
    type: Ready
```
[2]
```
Exception: Could not retrieve schema from tcp:ovsdbserver-sb-0.openstack.svc.cluster.local:6642
```

[1] https://logserver.rdoproject.org/17/217/d929c50b05f7ad4137efda0dccbf8b49b1e75d5f/github-check/ovn-operator-tempest-multinode/1e0c3ef/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/crs/openstackcontrolplanes.core.openstack.org/openstack-galera-network-isolation.yaml

[2]https://logserver.rdoproject.org/17/217/d929c50b05f7ad4137efda0dccbf8b49b1e75d5f/github-check/ovn-operator-tempest-multinode/1e0c3ef/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/neutron-7cfb6b45dd-pd95b/neutron-7cfb6b45dd-pd95b-describe

**Comment by lewisdenny** [comment]
recheck

File: controllers/ovndbcluster_controller.go [hunk 2]
```diff
@@ -533,9 +533,16 @@ func (r *OVNDBClusterReconciler) reconcileServices(
 	//
 	// Ensure the ovndbcluster headless service Exists
 	//
+	headlessServiceLabels := map[string]string{
```
Comment (karelyatin) [inline controllers/ovndbcluster_controller.go:536]: can just extend serviceLabels here instead of duplicating.

File: controllers/ovndbcluster_controller.go [hunk 3]
```diff
@@ -533,9 +533,16 @@ func (r *OVNDBClusterReconciler) reconcileServices(
...
+		common.AppSelector: serviceName,
+		"type":             "headless",
+	}
+	headlessSelectorLabels := map[string]string{
```
Comment (karelyatin) [inline controllers/ovndbcluster_controller.go:540]: can use serviceLabels instead of defining new variable here.

File: controllers/ovndbcluster_controller.go [hunk 4]
```diff
@@ -562,9 +569,14 @@ func (r *OVNDBClusterReconciler) reconcileServices(
 		ovndbServiceLabels := map[string]string{
 			common.AppSelector:                   serviceName,
```
Comment (karelyatin) [inline controllers/ovndbcluster_controller.go:570]: same here, can extend ovndbSelectorLabels

File: controllers/ovndbcluster_controller.go [hunk 5]
```diff
@@ -581,11 +593,15 @@ func (r *OVNDBClusterReconciler) reconcileServices(
 	}
 
 	// Delete any extra services left after scale down
+	normalServiceLabels := map[string]string{
```
Comment (karelyatin) [inline controllers/ovndbcluster_controller.go:596]: normal vs cluster in name. wdyt to use"regular" as type?

Comment (averdagu) [reply controllers/ovndbcluster_controller.go:596]: The problem with "normal" or "regular" is that can be missleading, as it doesn't mean anything inside kubernetes, what's normal or regular? That's why I changed for Cluster (As one of the differences bw headless services and non headless services is that one has Cluster Ip and the other has not). That's why "type: cluster" is the best name I found so far). wdyt?

Also, true, will change normalServiceLabels to clusterServiceLabels, as there's a mismatch on naming :)

File: controllers/ovndbcluster_controller.go (hunk 2 above)
Comment (averdagu) [reply controllers/ovndbcluster_controller.go:536]: Extending map from serviceLabels would modify servicelabels, we could use maps.Clone https://pkg.go.dev/maps

Comment (karelyatin) [reply controllers/ovndbcluster_controller.go:536]: i was considering like using something ilke https://github.com/openstack-k8s-operators/lib-common/blob/c9467a8fbbfc3e244207cd3d435975aca298d33d/modules/common/util/map.go#L86

File: controllers/ovndbcluster_controller.go [hunk 6]
```diff
@@ -533,9 +533,10 @@ func (r *OVNDBClusterReconciler) reconcileServices(
 	//
 	// Ensure the ovndbcluster headless service Exists
 	//
+	headlessServiceLabels := util.MergeMaps(serviceLabels, map[string]string{"type": "cluster"})
```
Comment (karelyatin) [inline controllers/ovndbcluster_controller.go:536]: type: headless

File: controllers/ovndbcluster_controller.go [hunk 7]
```diff
@@ -581,11 +583,15 @@ func (r *OVNDBClusterReconciler) reconcileServices(
 	}
 
 	// Delete any extra services left after scale down
+	clusterServiceLabels := map[string]string{
```
Comment (karelyatin) [inline controllers/ovndbcluster_controller.go:586]: clusterServiceLabels := util.MergeMaps(serviceLabels, map[string]string{"type": "cluster"})

**Review by karelyatin — CHANGES_REQUESTED** [review]

File: controllers/ovndbcluster_controller.go (hunk 6 above)
Comment (booxter) [reply controllers/ovndbcluster_controller.go:536]: this should probably be a const in v1betav1/ module

**Review by booxter — CHANGES_REQUESTED** [review]
My main request here is that a test case should be added to demonstrate the failing scenario.

(It would also be nice to have some documentation as to which services are DNS records are created by the operator, because it becomes hard to follow the changes in the area. But this is not a requirement for this PR, just an observation that some docs are probably due and I would appreciate if you could propose some for this repo.)

File: controllers/ovndbcluster_controller.go (hunk 6 above)
Comment (averdagu) [reply controllers/ovndbcluster_controller.go:536]: Thanks Yatin and Ihar, will apply changes

File: controllers/ovndbcluster_controller.go (hunk 7 above)
Comment (averdagu) [reply controllers/ovndbcluster_controller.go:586]: Done

File: tests/functional/ovndbcluster_controller_test.go [hunk 8]
```diff
@@ -226,6 +227,51 @@ var _ = Describe("OVNDBCluster controller", func() {
...
+				statefulSetName,
+				map[string][]string{namespace + "/internalapi": {"10.0.0.1"}},
+			)
+			service := types.NamespacedName{
+				Namespace: OVNDBClusterName.Namespace,
+				Name:      "ovsdbserver-sb",
+			}
+			// Even with the Eventually OVN takes to long to reconcile and create both services
+			// with this extra 2 seconds by the time the check is done, services will be already created
+			time.Sleep(2 * time.Second)
```
Comment (karelyatin) [inline tests/functional/ovndbcluster_controller_test.go:249]: this doesn't seem needed, Can you try dropping this workaround + fixing operator-lint issues

File: tests/functional/ovndbcluster_controller_test.go [hunk 9]
```diff
@@ -226,6 +227,51 @@ var _ = Describe("OVNDBCluster controller", func() {
...
+
+			statefulSetName := types.NamespacedName{
+				Namespace: namespace,
+				Name:      "ovsdbserver-sb",
+			}
+			th.SimulateStatefulSetReplicaReadyWithPods(
+				statefulSetName,
+				map[string][]string{namespace + "/internalapi": {"10.0.0.1"}},
+			)
+			service := types.NamespacedName{
```
Comment (karelyatin) [inline tests/functional/ovndbcluster_controller_test.go:243]: namespacename object doesn't look necessary, can just pass the namespace as that's what used by GetServicesListWithLabel.

File: tests/functional/ovndbcluster_controller_test.go [hunk 10]
```diff
@@ -226,6 +227,51 @@ var _ = Describe("OVNDBCluster controller", func() {
...
+			// Cluster is created with 1 replica, serviceListWithoutTypeLabel should be 2:
+			// - ovsdbserver-sb   (headless type)
+			// - ovsdbserver-sb-0 (cluster type)
+			Eventually(func(g Gomega) {
+				serviceListWithoutTypeLabel := GetServicesListWithLabel(service)
+				Expect(len(serviceListWithoutTypeLabel.Items)).To(Equal(2))
+			}).Should(Succeed())
+
+			Eventually(func(g Gomega) {
+				clusterService := types.NamespacedName{
```
Comment (karelyatin) [inline tests/functional/ovndbcluster_controller_test.go:259]: same

File: tests/functional/ovndbcluster_controller_test.go (hunk 9 above)
Comment (averdagu) [reply tests/functional/ovndbcluster_controller_test.go:243]: Done

File: tests/functional/ovndbcluster_controller_test.go (hunk 10 above)
Comment (averdagu) [reply tests/functional/ovndbcluster_controller_test.go:259]: Done

File: tests/functional/ovndbcluster_controller_test.go (hunk 8 above)
Comment (averdagu) [reply tests/functional/ovndbcluster_controller_test.go]: Done

**Comment by averdagu** [comment]
recheck

**Review by karelyatin — APPROVED** [review]
